doc/content/design/snapshot-revert.md

```diff
@@ -1,85 +1,102 @@
 ---
-title: Improving snapshot revert behaviour
+title: Better VM revert
 layout: default
 design_doc: true
-revision: 1
+revision: 2
 status: confirmed
 ---
 
-Currently there is a XenAPI `VM.revert` which reverts a "VM" to the state it
-was in when a VM-level snapshot was taken. There is no `VDI.revert` so
-`VM.revert` uses `VDI.clone` to change the state of the disks.
+## Overview
 
-The use of `VDI.clone` has the side-effect of changing VDI refs and uuids.
-This causes the following problems:
+Currently there is a XenAPI `VM.revert` which reverts a VM to the state it was
+in when a VM-level snapshot was taken. Because there is no `VDI.revert`,
+`VM.revert` uses `VDI.clone` on the snapshot to change the state of the disks.
 
-- It is difficult for clients
-  such as [Apache CloudStack](http://cloudstack.apache.org) to keep track
-  of the disks it is actively managing
-- VDI snapshot metadata (`VDI.snapshot_of` et al) has to be carefully
-  fixed up since all the old refs are now dangling
+Because `VDI.clone` creates new VDI refs and uuids, some problematic
+behaviours arise:
+
+- Clients such as 
+  [Apache CloudStack](http://cloudstack.apache.org) need to include complex
+  logic to keep track of the disks they are actively managing
+- Because the snapshot is cloned and the original vdi is deleted, VDI
+  references to the VDI become invalid, like `VDI.snapshot_of`. This means
+  there the database has to be combed through to change these references. 
+  Because the database doesn't support transaction this operation is not atomic
+  and can produce inconsistent database states.
 
 We will fix these problems by:
 
-1. adding a `VDI.revert` to the SMAPIv2 and calling this from `VM.revert`
-2. defining a new SMAPIv1 operation `vdi_revert` and a corresponding capability
-   `VDI_REVERT`
-3. the Xapi implementation of `VDI.revert` will first try the `vdi_revert`,
-   and fall back to `VDI.clone` if that fails
-4. implement `vdi_revert` for common storage types, including File and LVM-based
-   SRs.
+- defining a new storage operation `VDI.revert` in storage_interface and
+  calling this from `VM.revert`
+- proxying the storage operation to SMAPIv3 and SMAPv1 backends accordingly
+- changing the Xapi implementation of `VM.revert` to first try the
```

Comment:
This should be exposed as a feature the SM layer can expose when supported and then there is no need to try with failback.